### PR TITLE
Mj reintroduce sepa

### PR DESCRIPTION
**Pull request by michaelbjacobson**
## What are you doing in this PR?
[**Trello Card**](https://trello.com/c/TlfxGmOx/189-re-enable-sepa-on-the-contributions-checkout)

This updates our SEPA implementation to add `country` and `streetName` fields to the payment method. These fields are required for a certain set of country IBANs but, due to an ongoing issue with Zuora, we are going to require these fields for _all_ SEPA recurring contributions for the time being.

Alongside this work I planned to update the Braze email template for payment confirmation, to include the address fields when they are present, but this is not something we show on Direct Debit email confirmations, so I have not made this change.

## Is this an AB test?
- [ ] Yes
- [x] No

## Screenshots
Screenshot 2022-03-16 at 14 39 14

### CODE REVIEW

File: support-frontend/app/utils/CheckoutValidationRules.scala [hunk 1]
```diff
@@ -84,7 +84,7 @@ object PaidProductValidation {
       existingDetails.billingAccountId.nonEmpty.otherwise("existing billing account id missing")
     case AmazonPayPaymentFields(amazonPayBillingAgreementId) =>
       amazonPayBillingAgreementId.nonEmpty.otherwise("amazonPayBillingAgreementId missing")
-    case SepaPaymentFields(accountHolderName, iban) =>
+    case SepaPaymentFields(accountHolderName, iban, country, streetName) =>
```
Comment (michaelbjacobson) [inline support-frontend/app/utils/CheckoutValidationRules.scala:87]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/components/svgs/sepa.tsx [hunk 2]
```diff
@@ -8,7 +8,7 @@ export default function SvgSepa(): JSX.Element {
 			className="svg-sepa"
 			aria-hidden="true"
 			focusable="false"
-			fill="#052962"
+			fill="#272727"
```
Comment (michaelbjacobson) [inline support-frontend/assets/components/svgs/sepa.tsx:11]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/ContributionForm.tsx [hunk 3]
```diff
@@ -198,6 +203,12 @@ const mapDispatchToProps = (dispatch: (...args: any[]) => any) => ({
 	setSepaAccountHolderName: (name: string) => {
 		dispatch(setSepaAccountHolderName(name));
 	},
+	setSepaAddressStreetName: (addressStreetName: string) => {
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/ContributionForm.tsx:206]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx [hunk 4]
```diff
@@ -1,7 +1,16 @@
 import { css } from '@emotion/react';
+import type { Country } from '@guardian/consent-management-platform/dist/types/countries';
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:2]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/ContributionForm.tsx [hunk 5]
```diff
@@ -387,8 +400,12 @@ function ContributionForm(props: PropTypes): JSX.Element {
 						<SepaForm
 							iban={props.sepaData.iban}
 							accountHolderName={props.sepaData.accountHolderName}
+							addressStreetName={props.sepaData.streetName}
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/ContributionForm.tsx:403]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx [hunk 6]
```diff
@@ -17,36 +26,78 @@ const fieldsContainerStyles = css`
 		margin-top: ${space[3]}px;
 	}
 `;
+
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:29]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx [hunk 7]
```diff
@@ -17,36 +26,78 @@ const fieldsContainerStyles = css`
...
 	accountHolderName,
+	addressStreetName,
+	addressCountry,
+	updateAddressStreetName,
+	updateAddressCountry,
 	updateIban,
 	updateAccountHolderName,
 	checkoutFormHasBeenSubmitted,
 }: SepaFormProps) {
+	const [requireAddress, setRequireAddress] = useState(true);
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:53]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx [hunk 8]
```diff
@@ -55,19 +106,60 @@ export function SepaForm({
...
+						onChange={(e) => ibanOnChange(e.target.value)}
 						error={
 							checkoutFormHasBeenSubmitted && !isValidIban(iban)
 								? 'Please provide a valid IBAN'
-								: null
+								: undefined
 						}
 					/>
 				</div>
+
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:124]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/contributionsLandingReducer.ts [hunk 9]
```diff
@@ -76,6 +77,8 @@ export interface PayPalData {
 export interface SepaData {
 	iban: string | null;
 	accountHolderName: string | null;
+	streetName?: string;
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/contributionsLandingReducer.ts:80]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/contributionsLandingReducer.ts [hunk 10]
```diff
@@ -394,6 +399,24 @@ function createFormReducer() {
 					},
 				};
 
+			case 'SET_SEPA_ADDRESS_STREET_NAME':
```
Comment (michaelbjacobson) [inline support-frontend/assets/pages/contributions-landing/contributionsLandingReducer.ts:402]: ⚠️ Actual file change! ⚠️

File: support-models/src/main/scala/com/gu/support/workers/PaymentFields.scala [hunk 11]
```diff
@@ -55,6 +56,8 @@ case class DirectDebitPaymentFields(
 case class SepaPaymentFields(
     accountHolderName: String,
     iban: String,
+    country: Option[String],
```
Comment (michaelbjacobson) [inline support-models/src/main/scala/com/gu/support/workers/PaymentFields.scala:59]: ⚠️ Actual file change! ⚠️

File: support-workers/src/main/scala/com/gu/support/workers/lambdas/CreatePaymentMethod.scala [hunk 12]
```diff
@@ -189,6 +189,8 @@ class CreatePaymentMethod(servicesProvider: ServiceProvider = ServiceProvider)
       SepaPaymentMethod(
         BankTransferAccountName = sepa.accountHolderName,
         BankTransferAccountNumber = sepa.iban,
+        Country = sepa.country,
```
Comment (michaelbjacobson) [inline support-workers/src/main/scala/com/gu/support/workers/lambdas/CreatePaymentMethod.scala:192]: ⚠️ Actual file change! ⚠️

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx [hunk 13]
```diff
@@ -17,36 +26,78 @@ const fieldsContainerStyles = css`
...
+		// 	'PM',
+		// 	'SM',
+		// 	'CH',
+		// 	'WF',
+		// ];
+		//
+		// return requiredAddressPrefixes.includes(iban.substring(0, 2).toUpperCase());
+
+		return true;
+	};
```
Comment (TomPretty) [inline support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:77]: should we delete this for now?

File: support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts [hunk 14]
```diff
@@ -63,10 +63,6 @@ export type Action =
 			type: 'SET_ERROR';
 			error: Option<string>;
 	  }
-	| {
-			type: 'SET_STAGE';
-			stage: Stage;
-	  }
```
Comment (TomPretty) [inline support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts:69]: was this unused?

File: support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx (hunk 13 above)
Comment (michaelbjacobson) [reply support-frontend/assets/pages/contributions-landing/components/SepaForm.tsx:77]: Could do. I just thought it was worth keeping so we could reinstate it easily. I'm easy either way.

File: support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts (hunk 14 above)
Comment (michaelbjacobson) [reply support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts:69]: Yeah, it's a duplicate declaration.

Comment (michaelbjacobson) [reply support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts:69]: @TomPretty the remaining declaration is on line 50

Comment (TomPretty) [reply support-frontend/assets/pages/subscriptions-redemption/subscriptionsRedemptionReducer.ts:69]: nice!

**Review by rupertbates — APPROVED** [review]
Looks good to me, next time I'd put all the unrelated changes into a separate PR though.

**Comment by michaelbjacobson** [comment]
> Looks good to me, next time I'd put all the unrelated changes into a separate PR though.

Yeah, they all came in via IntelliJ trying to be clever with code quality checking etc. I'll avoid that next time!

**Pull request merged**